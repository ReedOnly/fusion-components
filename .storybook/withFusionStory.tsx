import React from 'react';
import Header from '../src/components/core/Header';
import Content from '../src/components/core/Content';
import Overlay from '../src/components/general/ApplicationGuidance/components/Overlay';

export default (title: string, padding: number = 16) => (stories: () => React.ReactNode) => {
    return (
        <>
            <Overlay fixed>
                <Header aside={null} start={null} content={null} />
            </Overlay>
            <Content>
                <Overlay>
                    <div
                        style={{
                            padding,
                            minHeight: '100%',
                            width: '100%',
                            maxWidth: '100%',
                            boxSizing: 'border-box',
                        }}
                    >
                        {stories()}
                    </div>
                </Overlay>
            </Content>
        </>
    );
};
