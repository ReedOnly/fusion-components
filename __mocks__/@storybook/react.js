import renderer from 'react-test-renderer';
import addons, { mockChannel } from '@storybook/addons';
import { withFusionContext } from '../../.storybook/withFusionContext';

addons.setChannel(mockChannel());

export const action = (actionName) => jest.fn();
const fusionContext = withFusionContext();
export const storiesOf = (groupName) => {
    const api = {
        decorators: [],
        parameters: {},

        add(storyName, story) {
            if (api.parameters.skipTest) {
                return;
            }

            describe(groupName, () => {
                it(storyName, () => {
                    // const decoratedStory = api.decorators.reduce((s, decorator) => {
                    //     try {
                    //         return decorator(s, { parameters: api.parameters });
                    //     } catch (e) {
                    //         return s;
                    //     }
                    // }, story);

                    const component = renderer.create(fusionContext(story));

                    // @TODO Enable custom names once released:
                    // > https://github.com/facebook/jest/pull/2094
                    expect(component.toJSON())
                        .toMatchSnapshot
                        // `${groupName}.${storyName}`
                        ();
                });
            });

            return api;
        },

        addDecorator() {
            return api;
        },

        addParameters(parameters) {
            api.parameters = { ...api.parameters, ...parameters };
            return api;
        },
    };

    return api;
};

export const addDecorator = () => {};
