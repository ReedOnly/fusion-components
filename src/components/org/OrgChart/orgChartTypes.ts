import { ReactNode } from 'react';

export type OrgChartItemProps<T> = {
    item: T;
} 

export type OrgChartProps<T> = {
    structure: T[];
    component: React.FC<OrgChartItemProps<T>> | null
    cardWidth?: number;
    cardHeight?: number;
    cardMargin?: number;
    rowMargin?: number;
};

export type OrgStructure = {
    id: string;
    parentId?: string;
    aside?: boolean;
}

export type OrgNode<T> = OrgStructure & {
    x:number;
    y:number;
    data: T;
}