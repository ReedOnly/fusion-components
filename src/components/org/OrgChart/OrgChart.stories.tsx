import React from 'react';
import { storiesOf } from '@storybook/react';
import withFusionStory from '../../../../.storybook/withFusionStory';
import OrgChart from '.';
import { OrgStructure, OrgChartItemProps } from './orgChartTypes';

type Position = OrgStructure & {
    name?: string;
    positionId?: string;
    personName?: string;
};

const ParentLinkComponent = () => <div> To parent</div>;

const positions: Position[] = [
    {
        id: '1',
        name: 'Offshore Worker',
        positionId: '123',
        personName: 'James Test',
        parentLink: ParentLinkComponent,
    },
    {
        id: '2',
        parentId: '1',
        name: 'Engineer',
        positionId: '123',
        personName: 'Sam Test',
        aside: true,
    },
    {
        id: '5',
        parentId: '1',
        name: 'Worker',
        positionId: '3',
        personName: 'Sam Test',
        aside: true,
    },
    {
        id: '2132',
        parentId: '1',
        name: 'Engineer',
        positionId: '123',
        personName: 'Sam Test',
        aside: true,
    },
    {
        id: '51245123',
        parentId: '1',
        name: 'Worker',
        positionId: '3',
        personName: 'Sam Test',
        aside: true,
    },
    {
        id: '3',
        parentId: '1',
        name: 'Offshore Worker',
        positionId: '3',
        personName: 'Jane Test',
    },
    {
        id: '32',
        parentId: '1',
        name: 'Offshore Worker',
        positionId: '3',
        personName: 'Test 123',
    },
    {
        id: '44',
        parentId: '1',
        name: 'Test worker 1',
        positionId: '4',
        personName: 'Test Test',
    },
    {
        id: '3123',
        parentId: '1',
        name: 'Offshore Worker',
        positionId: '3',
        personName: 'Jane Test',
    },
    {
        id: '3232',
        parentId: '1',
        name: 'Offshore Worker',
        positionId: '3',
        personName: 'Test 123',
    },
    {
        id: '44312',
        parentId: '1',
        name: 'Test worker 1',
        positionId: '4',
        personName: 'Test Test',
    },
    {
        id: '312123',
        parentId: '1',
        name: 'Offshore Worker',
        positionId: '3',
        personName: 'Jane Test',
    },
    {
        id: '323232',
        parentId: '1',
        name: 'Offshore Worker',
        positionId: '3',
        personName: 'Test 123',
    },
    {
        id: '4431212',
        parentId: '1',
        name: 'Test worker 1',
        positionId: '4',
        personName: 'Test Test',
    },
];

const PositionCard: React.FC<OrgChartItemProps<Position>> = ({ item }) => {
    return (
        <div
            style={{
                height: 'calc(100% - 20px)',
                display: 'flex',
                justifyContent: 'center',
                alignItems: 'center',
                backgroundColor: 'white',
                boxShadow:
                    '0px 1px 5px rgba(0, 0, 0, 0.2), 0px 3px 4px rgba(0, 0, 0, 0.12), 0px 2px 4px rgba(0, 0, 0, 0.14)',
                margin: '10px',
                boxSizing: 'border-box',
                flex: '1',
                fontSize: '16px',
            }}
        >
            Position : {item.name}
            <br />
            Person : {item.personName}
        </div>
    );
};

const OrgChartStory = () => {
    return (
        <div style={{ width: '100%', height: '100%' }}>
            <OrgChart
                structure={positions}
                component={PositionCard}
                asideLabel="ASIDE"
                childrenLabel="CHILDREN"
            />
        </div>
    );
};

storiesOf('Org|Org Chart', module)
    .addDecorator(withFusionStory('Org Chart'))
    .add('Default', () => <OrgChartStory />);
