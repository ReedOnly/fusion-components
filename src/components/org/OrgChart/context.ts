import React from 'react';
import { OrgNode, OrgChartItemProps } from './orgChartTypes';

export type OrgChartContextType<T> = {
    width: number;
    height: number;
    centerX: number;
    centerY: number;
    cardWidth: number;
    cardHeight: number;
    cardMargin: number;
    rowMargin: number;
    allNodes: OrgNode<T>[];
    component: React.FC<OrgChartItemProps<T>>
    asideRows: number;
    childrenRows: number;
    updateAsideRows: (rows: number) => void;
    updateChildrenRows: (rows: number) => void;
    updatePosition: (node: OrgNode<T>, x: number, y: number) => void;
};

export const OrgChartContext: React.Context<OrgChartContextType<any>> = React.createContext<OrgChartContextType<any>>({} as OrgChartContextType<any>);

