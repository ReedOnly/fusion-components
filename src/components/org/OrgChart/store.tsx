import { OrgNode, OrgChartItemProps } from './orgChartTypes';
import React, { FC, useReducer, Reducer, Context, createContext, Dispatch } from 'react';

type Action<T> =
    | { type: 'UPDATE_SIZE'; width?: number; height?: number }
    | { type: 'UPDATE_CENTER'; x?: number; y?: number }
    | { type: 'UPDATE_CARD_SIZE'; width?: number; height?: number; margin?: number }
    | { type: 'UPDATE_ROW_MARGIN'; margin: number }
    | { type: 'UPDATE_NODES'; nodes: OrgNode<T>[] }
    | { type: 'UPDATE_COMPONENT'; component: FC<OrgChartItemProps<T>> }
    | { type: 'UPDATE_ASIDE_ROWS'; rows: number }
    | { type: 'UPDATE_CHILDREN_ROWS'; rows: number }
    | { type: 'UPDATE_POSITION'; node: OrgNode<T>; x: number; y: number };

export type OrgChartContextType<T> = {
    width: number;
    height: number;
    centerX: number;
    centerY: number;
    cardWidth: number;
    cardHeight: number;
    cardMargin: number;
    rowMargin: number;
    allNodes: OrgNode<T>[];
    component: React.FC<OrgChartItemProps<T>> | null;
    asideRows: number;
    childrenRows: number;
};

export type OrgChartContextReducer<T> = {
    state: OrgChartContextType<T>;
    dispatch: Dispatch<Action<T>>;
};

export const OrgChartContext: Context<OrgChartContextReducer<any>> = createContext<
    OrgChartContextReducer<any>
>({} as OrgChartContextReducer<any>);

function reducer<T>(state: OrgChartContextType<T>, action: Action<T>): OrgChartContextType<T> {
    switch (action.type) {
        case 'UPDATE_SIZE':
            return {
                ...state,
                width: action.width || state.width,
                height: action.height || state.height,
            };
        case 'UPDATE_CENTER':
            return {
                ...state,
                centerX: action.x || state.centerX,
                centerY: action.x || state.centerY,
            };
        case 'UPDATE_CARD_SIZE':
            return {
                ...state,
                cardWidth: action.width || state.cardWidth,
                cardHeight: action.height || state.cardHeight,
                cardMargin: action.margin || state.cardMargin,
            };
        case 'UPDATE_ROW_MARGIN':
            return {
                ...state,
                rowMargin: action.margin,
            };
        case 'UPDATE_NODES':
            return {
                ...state,
                allNodes: action.nodes,
            };
        case 'UPDATE_COMPONENT':
            return {
                ...state,
                component: action.component,
            };
        case 'UPDATE_ASIDE_ROWS':
            return {
                ...state,
                asideRows: action.rows,
            };
        case 'UPDATE_CHILDREN_ROWS':
            return {
                ...state,
                childrenRows: action.rows,
            };
        case 'UPDATE_POSITION':
            return {
                ...state,
                allNodes: state.allNodes.map(prevNode => {
                    if (prevNode.id === action.node.id) {
                        return {
                            ...action.node,
                            x: action.x,
                            y: action.y,
                        };
                    } else {
                        return prevNode;
                    }
                }),
            };
    }
}

export function OrgChartContextProvider<T>({children}: any) {
    const initialState: OrgChartContextType<T> = {
        width: 0,
        height: 0,
        centerX: 0,
        centerY: 0,
        cardWidth: 0,
        cardHeight: 0,
        cardMargin: 0,
        rowMargin: 0,
        allNodes: [],
        asideRows: 0,
        childrenRows: 0,
        component: null,
    };

    const [state, dispatch] = useReducer<Reducer<OrgChartContextType<T>, Action<T>>>(
        reducer,
        initialState
    );
    const value = { state, dispatch };

    return <OrgChartContext.Provider value={value}>{children}</OrgChartContext.Provider>;
}
