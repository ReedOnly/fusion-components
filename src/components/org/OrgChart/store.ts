import { OrgChartContextType } from "./context";




export function OrgChartContextProvider<T>(children:any){

    const initialState: OrgChartContextType<T> = {
        width: 0,
        height: 0,
        centerX: 0,
        centerY: 0,
        cardWidth: 0,
        cardHeight: 0,
        cardMargin: 0,
        rowMargin: 0,
        allNodes: [],
        asideRows: 0,
        childrenRows: 0,
    }
}