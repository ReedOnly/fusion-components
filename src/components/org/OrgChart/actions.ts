import { OrgNode, OrgChartProps, OrgStructure } from './orgChartTypes';
import { useEffect, useContext } from 'react';
import { OrgChartContextReducer, OrgChartContext } from './store';

type OrgChartActionProps<T> = OrgChartProps<T> & {
    parentHeight: number;
    parentWidth: number;
};

const useOrgChartActions = <T extends OrgStructure>({
    structure,
    cardWidth = 320,
    cardHeight = 112,
    cardMargin = 16,
    rowMargin = 128,
    component,
    childrenLabel,
    asideLabel,
    breadCrumbComponent,
    parentHeight,
    parentWidth,
    breadCrumbs,
}: OrgChartActionProps<T>) => {
    const { dispatch } = useContext<OrgChartContextReducer<T>>(OrgChartContext);

    const generateNodes = (structure: T[]): OrgNode<T>[] => {
        return structure.map(item => {
            return {
                id: item.id,
                parentId: item.parentId,
                x: 0,
                y: 0,
                data: item,
                aside: item.aside,
            };
        });
    };

    useEffect(() => {
        dispatch({
            type: 'UPDATE_LABELS',
            asideLabel: asideLabel,
            childrenLabel: childrenLabel,
        });
    }, [asideLabel, childrenLabel]);

    useEffect(() => {
        dispatch({
            type: 'UPDATE_NODES',
            nodes: generateNodes(structure),
        });
    }, [structure]);

    useEffect(() => {
        if (component || breadCrumbComponent) {
            dispatch({
                type: 'UPDATE_COMPONENTS',
                component: component || undefined,
                breadCrumbComponent: breadCrumbComponent || undefined,
            });
        }
    }, [component, breadCrumbComponent]);

    useEffect(() => {
        dispatch({
            type: 'UPDATE_CARD_SIZE',
            height: cardHeight,
            width: cardWidth,
            margin: cardMargin,
            initialWidth: cardWidth,
        });
    }, [cardHeight, cardWidth, cardMargin]);

    useEffect(() => {
        if (rowMargin) {
            dispatch({
                type: 'UPDATE_ROW_MARGIN',
                margin: rowMargin,
            });
        }
    }, [rowMargin]);

    useEffect(() => {
        dispatch({
            type: 'UPDATE_SIZE',
            height: parentHeight,
            width: parentWidth,
        });
        dispatch({
            type: 'UPDATE_CENTER',
            x: parentWidth / 2,
            y: parentHeight / 2,
        });
    }, [parentHeight, parentWidth]);

    useEffect(() => {
            dispatch({
                type: 'UPDATE_BREADCRUMBS',
                breadcrumbs: breadCrumbs || null,
            });
    }, [breadCrumbs]);
};

export default useOrgChartActions;
