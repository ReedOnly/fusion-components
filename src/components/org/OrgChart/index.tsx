import React, { useRef, useMemo, useContext } from 'react';
import useParentSize from './hooks/useParentSize';
import { OrgChartContextProvider, OrgChartContextReducer, OrgChartContext } from './store';
import { OrgStructure, OrgChartProps, OrgChartItemProps } from './orgChartTypes';
import Links from './components/Links';
import Root from './components/Root';
import Aside from './components/Aside';
import Children from './components/Children';
import Labels from './components/Labels';
import useOrgChartActions from './actions';
import BreadCrumbs from './components/BreadCrumbs';

export { OrgStructure, OrgChartItemProps };

const OrgChart = <T extends OrgStructure>(props: OrgChartProps<T>) => (
    <OrgChartContextProvider>
        <OrgChartContent {...props} />
    </OrgChartContextProvider>
);

const OrgChartContent = <T extends OrgStructure>(props: OrgChartProps<T>) => {
    const orgContainerRef = useRef<SVGSVGElement | null>(null);
    const { height, width } = useParentSize(orgContainerRef);

    useOrgChartActions({ ...props, parentHeight: height, parentWidth: width });

    const {
        state: { rowMargin, asideRows, childrenRows },
    } = useContext<OrgChartContextReducer<T>>(OrgChartContext);

    const svgHeight = useMemo(() => {
        const rootMargin = rowMargin;
        const labelMargin = 60;
        const asideMargin = rowMargin * asideRows;
        const childrenMargin = rowMargin * childrenRows;
        return asideMargin + childrenMargin + rootMargin + labelMargin;
    }, [rowMargin, asideRows, childrenRows]);

    return (
        <svg
            ref={orgContainerRef}
            width={width}
            height={svgHeight}
            viewBox={`0 0 ${width} ${svgHeight}`}
        >
            <Links />
            <Root />
            <Aside />
            <Children />
            <Labels />
            <BreadCrumbs />
        </svg>
    );
};

export default OrgChart;
