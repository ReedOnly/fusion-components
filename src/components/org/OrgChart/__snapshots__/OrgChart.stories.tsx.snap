// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pro org|Org Chart Default 1`] = `
<div
  className="container comfortable"
>
  <div
    className="container comfortable"
  >
    <div
      style={
        Object {
          "height": "100%",
          "width": "100%",
        }
      }
    >
      <svg
        height={60}
        viewBox="0 0 0 60"
        width={0}
      >
        <g
          className="links"
        />
        <g
          className="aside"
        />
        <g
          className="root"
        />
        <g
          className="children"
        />
        <g
          className="label"
        >
          0
          0
        </g>
      </svg>
    </div>
  </div>
  <div
    className="overlay"
  />
  <div
    className="snacks"
  />
  <div
    className="dialog"
  />
</div>
`;
